apps/heft/src/cli/actions/TestAction.ts

```diff
@@ -1,40 +1,115 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
 
-import { SyncHook } from 'tapable';
+import { SyncHook, AsyncParallelHook, AsyncSeriesHook } from 'tapable';
+import { CommandLineFlagParameter } from '@rushstack/ts-command-line';
 
 import { BuildAction, IBuildActionOptions, IBuildActionContext } from './BuildAction';
 import { ActionHooksBase, IActionContext } from './HeftActionBase';
 
 /**
  * @public
  */
-export class TestHooks extends ActionHooksBase<ITestActionProperties> {}
+export class TestHooks extends ActionHooksBase<ITestActionProperties> {
+  public readonly run: AsyncParallelHook = new AsyncParallelHook();
+  public readonly configureTest: AsyncSeriesHook = new AsyncSeriesHook();
+}
 
 /**
  * @public
  */
-export interface ITestActionProperties {}
+export interface ITestActionProperties {
+  watchMode: boolean;
+}
 
 /**
  * @public
  */
 export interface ITestActionContext extends IActionContext<TestHooks, ITestActionProperties> {}
 
-export interface ITestActionOptions extends IBuildActionOptions {}
+export interface ITestActionOptions extends IBuildActionOptions {
+  buildAction: BuildAction;
+}
 
 export class TestAction extends BuildAction {
   public testActionHook: SyncHook<ITestActionContext> = new SyncHook<ITestActionContext>(['action']);
+  private _buildAction: BuildAction;
+
+  private _noBuildFlag: CommandLineFlagParameter;
 
   public constructor(options: ITestActionOptions) {
     super(options, {
       actionName: 'test',
       summary: 'Build the project and run tests.',
       documentation: ''
     });
+
+    this._buildAction = options.buildAction;
+  }
+
+  public onDefineParameters(): void {
+    super.onDefineParameters();
+
+    this._noBuildFlag = this.defineFlagParameter({
+      parameterLongName: '--no-build',
+      description: 'If provided, only run tests. Do not build first.'
+    });
   }
 
   protected async actionExecute(buildActionContext: IBuildActionContext): Promise<void> {
-    throw new Error('Not implemented yet...');
+    const testActionContext: ITestActionContext = {
+      hooks: new TestHooks(),
+      properties: {
+        watchMode: buildActionContext.properties.watchMode
+      }
+    };
+    const shouldBuild: boolean = !this._noBuildFlag.value;
+
+    if (testActionContext.properties.watchMode) {
+      if (!shouldBuild) {
+        throw new Error(`${this._watchFlag.longName} is not compatible with ${this._noBuildFlag.longName}`);
+      } else if (buildActionContext.properties.noTest) {
```

Comment:
```
usage: heft test [-h] [-v] [--production] [--locale LOCALE] [-l] [--clean]
                 [--notest] [--max-old-space-size SIZE] [--watch]
                 [--typescript-max-write-parallelism PARALLEILSM] [--no-build]


Optional arguments:
  -h, --help            Show this help message and exit.
  -v, --verbose         If specified, log information useful for debugging.
  --production          If specified, build ship/production output
  --locale LOCALE       Only build the specified locale, if applicable.
  -l, --lite            Perform a minimal build, skipping optional steps like
                        linting.
  --clean               If specified, clean the package before rebuilding.
  --notest              If specified, run the build without testing.
  --max-old-space-size SIZE
                        Used to specify the max old space size.
  --watch               If provided, run tests in watch mode.
  --typescript-max-write-parallelism PARALLEILSM
                        Set this to change the maximum write parallelism.
                        This parameter overrides what is set in typescript.
                        json. The default is 50.
  --no-build            If provided, only run tests. Do not build first.
```

Is `heft test --notest` meaningful?  Also `--no-build`  has inconsistent punctuation.

Reply:
`hest test --notest` is currently in integration with BXL. Because Rush doesn't support multiple phases, a project's script block looks like this:

```
"scripts": {
  "build": "heft test",
  "test": "heft test --no-build"
}
```

BXL then runs the "build" script with the `--notest` parameter appended in one phase (called Pips in BXL parlance) and then runs the "test" script in a subsequent phase.

Check out the BXL docs here: https://github.com/microsoft/BuildXL/blob/93ac7b50826cab05a8ae18c9f899546c7f479590/Documentation/Wiki/Frontends/rush-commands.md#augmenting-the-script-command-arguments

These params definitely shouldn't exist long-term, but they're useful for the sake of ensuring that `rush build` and a BXL build have effectively the same behavior, but with better optimizations on the BXL side.